src/dscanner/analysis/assert_without_msg.d

```diff
@@ -5,132 +5,165 @@
+		super(fileName, skipTests);
+	
+		import std.file : readText;
+		this.code = readText(fileName);
 	}
 
-	override void visit(const AssertExpression expr)
+	mixin ScopedVisit!(AST.FuncDeclaration);
+	mixin ScopedVisit!(AST.IfStatement);
+	mixin ScopedVisit!(AST.WhileStatement);
+	mixin ScopedVisit!(AST.ForStatement);
+	mixin ScopedVisit!(AST.ForeachStatement);
+	mixin ScopedVisit!(AST.ClassDeclaration);
+	mixin ScopedVisit!(AST.StructDeclaration);
+	mixin ScopedVisit!(AST.InterfaceDeclaration);
+	mixin ScopedVisit!(AST.UnionDeclaration);
+	mixin ScopedVisit!(AST.ScopeStatement);
+
+	override void visit(AST.AssertExp ae)
 	{
-		if (expr.assertArguments && expr.assertArguments.message is null)
-			addErrorMessage(expr.line, expr.column, KEY, MESSAGE);
+		// We need to ensure that this is an actual `assert` expression,
+		// Not an in/out contract represented also as an AssertExp in the AST
+		import std.string : indexOf;
+
+		// File Offset indicates where the expression starts
+		// To extract the function name get rid of `(` and everything that follows
+		string funcName = code[ae.loc.fileOffset .. $];
+		funcName = funcName[0 .. funcName.indexOf("(")];
+
+		if (funcName == "assert" && !ae.msg)
+		// if (!ae.msg)
+			addErrorMessage(ae.loc.linnum, ae.loc.charnum, KEY, MESSAGE);
 	}
 
-	override void visit(const FunctionCallExpression expr)
+	override void visit(AST.StaticAssert ae)
 	{
-		if (!isStdExceptionImported)
+		if (!ae.msg)
+			addErrorMessage(ae.loc.linnum, ae.loc.charnum, KEY, MESSAGE);
+	}
+
+	override void visit(AST.Import i)
+	{
+		import std.algorithm.iteration : map;
+		import std.array : array;
+
+		const(char[])[] exceptionImportStr = ["std", "exception"];
+		
+		const(char[])[] aux = i.packages.map!(a => a.toString()).array;
+		aux ~= i.id.toString();
+
+		// import std.exception without selective imports
+		if (i.names.length == 0 && aux == exceptionImportStr)
+		{
+			isStdExceptionImported = true;
 			return;
+		}
 
-		if (const IdentifierOrTemplateInstance iot = safeAccess(expr)
-			.unaryExpression.primaryExpression.identifierOrTemplateInstance)
+		// Encountered std.exception with selective imports
+		foreach (selectiveImport; i.names)
 		{
-			auto ident = iot.identifier;
-			if (ident.text == "enforce" && expr.arguments !is null && expr.arguments.argumentList !is null &&
-					expr.arguments.argumentList.items.length < 2)
-				addErrorMessage(ident.line, ident.column, KEY, MESSAGE);
+			if (selectiveImport.toString() == "enforce")
+			{
+				isStdExceptionImported = true;
+				return;
+			}
 		}
 	}
 
-	override void visit(const SingleImport sImport)
+	// Visit ExpStatement instead of CallExp in order to
+	// have access to the location of the function identifier
+	override void visit(AST.ExpStatement es)
 	{
-		static immutable stdException = ["std", "exception"];
-		if (sImport.identifierChain.identifiers.map!(a => a.text).equal(stdException))
-			isStdExceptionImported = true;
-	}
+		auto ce = es.exp.isCallExp();
```

Comment:
I don't understand what extra information `ExpStatement` has that `CallExp` doesn't. The location of both should be identical. But still, I don't understand when why you have to look through the bytes to get the function name. All the information is already there in the AST. `CallExp.e1` should contain the name of the function

Reply:
Corresct, `ExpStatement` does not make sense here, the reason I used it is I think I found a bug regarding the location of a `CallExp`. I will need to look into that, but I agree with you